File: app/src/main/java/com/test/messages/demo/data/MessageDiffCallback.java
package com.test.messages.demo.data;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;

public class  MessageDiffCallback extends DiffUtil.ItemCallback<Message> {

    @Override
    public boolean areItemsTheSame(@NonNull Message oldItem, @NonNull Message newItem) {
        return oldItem.getId() == newItem.getId();
    }

    @Override
    public boolean areContentsTheSame(@NonNull Message oldItem, @NonNull Message newItem) {
        return oldItem.getSender().equals(newItem.getSender())
                && oldItem.getContent().equals(newItem.getContent());

    }

    @Override
    public Object getChangePayload(@NonNull Message oldItem, @NonNull Message newItem) {
        return null;
    }
}

File: app/src/main/java/com/test/messages/demo/data/MessageDiffCallback.kt
package com.test.messages.demo.data

import androidx.recyclerview.widget.DiffUtil

class MessageDiffCallback(
    private val oldList: List<MessageItem>,
    private val newList: List<MessageItem>
) : DiffUtil.Callback() {

    override fun getOldListSize(): Int = oldList.size

    override fun getNewListSize(): Int = newList.size

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        // Compare unique identifiers (e.g., threadId, or any unique field)
        return oldList[oldItemPosition].threadId == newList[newItemPosition].threadId
    }

    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        // Compare the contents of the item
        return oldList[oldItemPosition] == newList[newItemPosition]
    }
}